File: servers/cu_rs/ao-scheduler-utils/src/client/gateway.rs
use ao_common::models::gql_models::{Node, TransactionConnectionSchema};
use ao_common::models::shared_models::Tag;
use serde::Serialize;
use reqwest::Error;
use ao_common::arweave::InternalArweave;
use derive_more::{Display, Error};

use crate::err::InvalidSchedulerLocationError;

const URL_TAG: &str = "Url";
const TTL_TAG: &str = "Time-To-Live";
const _SCHEDULER_TAG: &str = "Scheduler";

pub struct Gateway {
    arweave: InternalArweave
}

impl Gateway {
    fn find_tag_value(name: &str, tags: &Vec<Tag>) -> String {
        match tags.iter().find(|tag| tag.name == name) {
            Some(found_tag) => found_tag.value.to_string(),
            None => "".to_string()
        }
    }
    
    fn find_transaction_tags<'a>(err_msg: &'a str, transaction: &'a Option<Node>) -> Result<Vec<Tag>, FindTxTagsError> {
        if let Some(tx) = transaction {
            return Ok(tx.tags.clone());
        }
        Err(FindTxTagsError { message: err_msg.to_string() })
    }
    
    async fn gateway_with<'a, T: Serialize, U: for<'de> serde::Deserialize<'de>>(&self, gateway_url: &'a str, query: &'a str, variables: T) -> Result<U, Error> {
        let result = self.arweave.query_gateway_with::<T, U>(gateway_url, query, variables).await;
    
        match result {
            Ok(res) => Ok(res),
            Err(e) => Err(e)
        }
    }

    pub async fn load_process_scheduler_with<'a>(&self, _gateway_url: &'a str) {

    }

    pub async fn load_scheduler_with<'a>(&self, gateway_url: &'a str, wallet_address: &'a str) -> Result<SchedulerResult, Box<dyn std::error::Error>> {
        #[allow(non_snake_case)]
        let GET_SCHEDULER_LOCATION = r#"
            query GetSchedulerLocation ($owner: String!) {
            transactions (
                owners: [$owner]
                tags: [
                { name: "Data-Protocol", values: ["ao"] },
                { name: "Type", values: ["Scheduler-Location"] }
                ]
                # Only need the most recent Scheduler-Location
                sort: HEIGHT_DESC
                first: 1
            ) {
                edges {
                node {
                    tags {
                    name
                    value
                    }
                }
                }
            }
            }
        "#;

        let result = self.gateway_with::<WalletAddress, TransactionConnectionSchema>(gateway_url, GET_SCHEDULER_LOCATION, WalletAddress { owner: wallet_address }).await;
        match result {
            Ok(tx) => {
                let tags = Gateway::find_transaction_tags(
                    format!("Could not find 'Scheduler-Location' owner by wallet {}", wallet_address).as_str(), 
                    &Some(tx.data.edges[0].node.clone())
                );
                match tags {
                    Ok(tags) => {
                        let url = Gateway::find_tag_value(URL_TAG, &tags);
                        let ttl = Gateway::find_tag_value(TTL_TAG, &tags);

                        if url.is_empty() {
                            let mut error = InvalidSchedulerLocationError::new();
                            error.message = "No 'Url' tag found on Scheduler-Location";
                            return Err(Box::new(error));
                        }
                        if ttl.is_empty() {
                            let mut error = InvalidSchedulerLocationError::new();
                            error.message = "No 'Time-To-Live' tag found on Scheduler-Location";
                            return Err(Box::new(error));
                        }
                        return Ok(SchedulerResult {
                            url,
                            ttl,
                            owner: wallet_address.to_string()
                        });
                    },
                    Err(e) => {
                        return Err(Box::new(e));
                    }
                }
            },
            Err(e) => Err(Box::new(e))
        }
    }
}

#[derive(Serialize)]
struct WalletAddress<'a> {
    owner: &'a str
}

#[derive(Error, Display, Debug)]
struct FindTxTagsError {
    message: String
}

pub struct SchedulerResult {
    pub url: String,
    pub ttl: String,
    pub owner: String
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_tag_value() {
        let tags = vec![
            Tag { name: "hello".to_string(), value: "22".to_string() },
            Tag { name: "world".to_string(), value: "tim".to_string() }
        ];

        let value = Gateway::find_tag_value("world", &tags);

        assert!(value != "22".to_string());
        assert!(value == "tim".to_string());
    }
}

File: scheduler-utils/src/client/gateway.js
import { defaultTo, find, juxt, path, pipe, prop } from 'ramda'

import { InvalidSchedulerLocationError, SchedulerTagNotFoundError, TransactionNotFoundError } from '../err.js'
import { backoff, okRes } from '../utils.js'

const URL_TAG = 'Url'
const TTL_TAG = 'Time-To-Live'
const SCHEDULER_TAG = 'Scheduler'

const findTagValue = (name) => pipe(
  defaultTo([]),
  find(tag => tag.name && tag.name.toLowerCase() === name.toLowerCase()),
  defaultTo({}),
  prop('value')
)

const findTransactionTags = (err) => pipe(
  (transaction) => {
    if (!transaction) throw new TransactionNotFoundError(err)
    return transaction
  },
  prop('tags'),
  defaultTo([])
)

export const parseHyperBeamResponse = (process) => {
  const commitments = process.commitments
  const processId = Object.keys(commitments).find(key => commitments[key].type === 'rsa-pss-sha256')

  if (!processId) {
    return { id: undefined, tags: [] }
  }

  const signedCommitment = commitments[processId]
  const committed = signedCommitment.committed
  const originalTags = Object.values(signedCommitment['original-tags'])
  const tags = []
  for (const tag of originalTags) {
    const { name, value } = tag
    if (committed.includes(name.toLowerCase())) {
      tags.push({ name, value })
    }
  }
  return { id: processId, tags }
}

function gatewayWith ({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES = 0, GRAPHQL_RETRY_BACKOFF = 300 }) {
  return async ({ query, variables }) => {
    return backoff(
      () => fetch(GRAPHQL_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ query, variables })
      })
        .then(okRes)
        .then((res) => res.json()),
      { maxRetries: GRAPHQL_MAX_RETRIES, delay: GRAPHQL_RETRY_BACKOFF })
  }
}

export function loadProcessWith ({ fetch, HB_GRAPHQL_URL, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF }) {
  const gateway = gatewayWith({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF })
  const GET_PROCESS_QUERY = `
        query ($ids: [ID!]!) {
        transactions(
          ids: $ids
          tags: [
            { name: "Type", values: ["Process"] }
        ]) {
          pageInfo {
            hasNextPage
          }
          edges {
            cursor
            node {
              id
              anchor
              signature
              recipient
              block {
                timestamp
              }
              owner {
                address
                key
              }
              fee {
                winston
                ar
              }
              quantity {
                winston
                ar
              }
              data {
                size
                type
              }
              tags {
                name
                value
              }
              block {
                id
                timestamp
                height
                previous
              }
              parent {
                id
              }
            }
          }
        }
      }
  `
  return async (process) => {
    return fetch(`${HB_GRAPHQL_URL}/${process}/serialize~json@1.0`, {
      headers: {
        Accept: 'application/json'
      }
    })
      .then((res) => res.json())
      .then(parseHyperBeamResponse)
      .catch(_e => {
        return gateway({ query: GET_PROCESS_QUERY, variables: { ids: [process] } })
          .then(path(['data', 'transactions', 'edges', '0', 'node']))
      })
  }
}

export function loadProcessSchedulerWith ({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF }) {
  const gateway = gatewayWith({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF })
  const loadScheduler = loadSchedulerWith({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF })

  const GET_TRANSACTIONS_QUERY = `
    query GetTransactions ($transactionIds: [ID!]!) {
      transactions(
        ids: $transactionIds
        tags: [
          { name: "Data-Protocol", values: ["ao"] }
        ]
      ) {
        edges {
          node {
            tags {
              name
              value
            }
          }
        }
      }
    }
  `

  const TESTNET_SU_ROUTER = 'https://su-router.ao-testnet.xyz'
  return async (process) => {
    /**
     * TODO: remove eagerly checking testnet su router once
     * gateway issues have been mitigated
     */
    const eagerTestnet = await fetch(`https://su-router.ao-testnet.xyz?process-id=${process}`)
      .then((res) => !res.ok
        ? undefined
        : res.json()
          .then(({ address }) => ({
            url: TESTNET_SU_ROUTER,
            ttl: 1000 * 60 * 60 * 48,
            address
          }))
      )
    if (eagerTestnet) return eagerTestnet

    return gateway({ query: GET_TRANSACTIONS_QUERY, variables: { transactionIds: [process] } })
      .then(path(['data', 'transactions', 'edges', '0', 'node']))
      .then(findTransactionTags(`Process ${process} was not found on gateway`))
      .then(findTagValue(SCHEDULER_TAG))
      .then((walletAddress) => {
        if (!walletAddress) throw new SchedulerTagNotFoundError('No "Scheduler" tag found on process')
        return loadScheduler(walletAddress)
      })
  }
}

export function loadSchedulerWith ({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF }) {
  const gateway = gatewayWith({ fetch, GRAPHQL_URL, GRAPHQL_MAX_RETRIES, GRAPHQL_RETRY_BACKOFF })

  const GET_SCHEDULER_LOCATION = `
    query GetSchedulerLocation ($owner: String!) {
      transactions (
        owners: [$owner]
        tags: [
          { name: "Data-Protocol", values: ["ao"] },
          { name: "Type", values: ["Scheduler-Location"] }
        ]
        # Only need the most recent Scheduler-Location
        sort: HEIGHT_DESC
        first: 1
      ) {
        edges {
          node {
            tags {
              name
              value
            }
          }
        }
      }
    }
  `

  return async (walletAddress) =>
    gateway({ query: GET_SCHEDULER_LOCATION, variables: { owner: walletAddress } })
      .then(path(['data', 'transactions', 'edges', '0', 'node']))
      .then(findTransactionTags(`Could not find 'Scheduler-Location' owner by wallet ${walletAddress}`))
      .then(juxt([
        findTagValue(URL_TAG),
        findTagValue(TTL_TAG)
      ]))
      .then(([url, ttl]) => {
        if (!url) throw new InvalidSchedulerLocationError('No "Url" tag found on Scheduler-Location')
        if (!ttl) throw new InvalidSchedulerLocationError('No "Time-To-Live" tag found on Scheduler-Location')
        return { url, ttl, address: walletAddress }
      })
}
